lib/cluster.py

```diff
@@ -300,3 +272,81 @@ def __repr__(self):
             s += str(lg)
         s += "\n"
         return s
+
+
+def discover_mode(conf_path):
+
+    with open(conf_path, "r") as config:
+        data = config.read()
+
+        # strip out templated variables {{ ... }} and sync functions `function ... }`
+        data = strip_invalid_json_from_config(data)
+
+        # Find all bucket names in config's databases: {}
+        conf_obj = json.loads(data)
+
+        # Add CBGT buckets
+        if "cluster_config" in conf_obj.keys():
+            return  "distributed_index"
+
+    return "channel_cache"
+    
+
+def strip_invalid_json_from_config(data):
+
+    # strip out templated variables {{ ... }}
+        data = re.sub("({{.*}})", "0", data)
+
+        # strip out sync functions `function ... }`
+        data = re.sub("(`function.*\n)(.*\n)+(.*}`)", "0", data)
+
+        return data
+
+def discover_buckets_from_config(conf_path, mode):
+
+    bucket_names_from_config = []
+
+    with open(conf_path, "r") as config:
+        data = config.read()
+
+        # strip out templated variables {{ ... }} and sync functions `function ... }`
+        data = strip_invalid_json_from_config(data)
+
+        # Find all bucket names in config's databases: {}
+        conf_obj = json.loads(data)
+
+        # Add CBGT buckets
+        if "cluster_config" in conf_obj.keys():
+            bucket_names_from_config.append(conf_obj["cluster_config"]["bucket"])
+
+        dbs = conf_obj["databases"]
+        for key, val in dbs.iteritems():
+            # Add data buckets
+            bucket_names_from_config.append(val["bucket"])
+            if "channel_index" in val:
+                # index buckets
+                bucket_names_from_config.append(val["channel_index"]["bucket"])
+
+            if not val.has_key("shadow"):
+                continue
+            
+            shadow = val["shadow"]
+            if len(shadow["bucket"]) > 0:
+                bucket_names_from_config.append(shadow["bucket"])
+
+    # Buckets may be shared for different functionality
+    bucket_name_set = list(set(bucket_names_from_config))
+
+    return bucket_name_set
+
+    
+if __name__ == "__main__":
+
+    # poor man's unit test
+    conf_path = "conf/sync_gateway_bucketshadow_cc.json"
```

Comment:
I think we should do this a bit different. Maybe we could return mode and buckets from the reset method. We may need to to that in the future anyway for config testing.